src/peft/tuners/bone/layer.py

```diff
@@ -0,0 +1,260 @@
+                    self.base_layer.weight.data += delta_weight
+                self.merged_adapters.append(active_adapter)
+
+    def unmerge(self) -> None:
+        """
+        This method unmerges all merged adapter layers from the base weights.
+        """
+        if not self.merged:
+            warnings.warn("Already unmerged. Nothing to do.")
+            return
+        while len(self.merged_adapters) > 0:
+            active_adapter = self.merged_adapters.pop()
+            if active_adapter in self.bone_block.keys():
+                orig_weight = self.get_base_layer().weight.data.clone()
+                delta_weight = self.get_delta_weight(active_adapter, orig_weight)
+                self.get_base_layer().weight.data -= delta_weight
+
+    def get_delta_weight(self, adapter, orig_weight) -> torch.Tensor:
+        """
+        Compute the delta weight for the given adapter.
+
+        Args:
+            adapter (str):
+                The name of the adapter for which the delta weight should be computed.
+        """
+        device = self.bone_block[adapter].device
+        dtype = self.bone_block[adapter].dtype
+        # In case users wants to merge the adapter weights that are in
+        # (b)float16 while being on CPU, we need to cast the weights to float32, perform the merge and then cast back to
+        # (b)float16 because some CPUs have slow bf16/fp16 matmuls.
+        cast_to_fp32 = device.type == "cpu" and (dtype == torch.float16 or dtype == torch.bfloat16)
+
+        weight_bone = self.bone_block[adapter]
+        
+        if cast_to_fp32:
+            weight_bone = weight_bone.float()
+        
+        #weight_bone = weight_bone.to(orig_weight.dtype)
+        w =  rearrange(orig_weight, '(a r1) (b r2) -> b a r1 r2', r1 = weight_bone.size(-1), r2 = weight_bone.size(-1))@weight_bone+weight_bone
+        output_tensor = rearrange(w, 'b a r1 r2 ->(a r1) (b r2) ')
+
+        if cast_to_fp32:
+            output_tensor = output_tensor.to(dtype=dtype)
+
+            # cast back the weights
+            self.bone_block[adapter].data = weight_bone.to(dtype)
+
+        return output_tensor
+
+    def forward(self, x: torch.Tensor, *args: Any, **kwargs: Any) -> torch.Tensor:
+        previous_dtype = x.dtype
+
+        if self.disable_adapters:
+            if self.merged:
+                self.unmerge()
+            result = self.base_layer(x, *args, **kwargs)
+        elif self.merged:
+            result = self.base_layer(x, *args, **kwargs)
+        else:
+            
+            for active_adapter in self.active_adapters:
+                if active_adapter not in self.bone_block.keys():
+                    continue
+                orig_weight = self.base_layer.weight.data.clone()
+                delta_weight = self.get_delta_weight(active_adapter, orig_weight)
+                orig_weight += delta_weight
+
+            result = F.linear(input=x, weight=orig_weight, bias=self.base_layer.bias)
```

Comment:
Just an idea from my side:

Right now, you're calculating `W2 = W + dW`, then `result = W2@x + bias`. Instead, I wonder if we could re-arrange this somehow to use `result = self.base_layer(x, *args, **kwargs)`, then `result = result + bone_result`. The `bone_result` would ideally be calculated in a parameter-efficient way, which should hopefully alleviate the memory overhead that Bone has compared to LoRA. For this, I wonder if the Bone contribution can be calculated group-wise and added incrementally instead of instantiating a big `dW`, so basically trading more compute for less memory. Do you think that would be possible?

Reply:
I think 'result = result + bone_result' directly is not feasible. The Bone repository has flash-bone that integrates linear and bone calculations, but I'm not very skilled at writing operators, so the current optimization speed is insufficient.

Reply:
If we have `Y = (W + W@dW + dW)@X + b`, we could rewrite it as `Y = W@X + (W@dW + dW)@X + b`, right? And since `W@X + b` is the base result, i.e. `self.base_layer(x, *args, **kwargs)`, this should work?

Another question would be if we can re-define `delta_weight` in a way that would allow to reduce the memory overhead.

https://github.com/huggingface/peft/pull/2148/files#diff-34b7ebce9fc88273adb34c46598bee8a92b72f5c1e7c420585e03938396a1448R210-R214

My thought was if we can somehow calculate the activations per block and add them incrementally to the base result. Maybe it's not possible or it would not help to reduce memory, but if we can trade off compute for less memory, it is usually a worthwhile trade. If it's simple not possible, feel free to ignore my comment :)

Reply:
Is it possible to add checkpointing specifically for bone_result? The optimization here indeed requires careful consideration.

Reply:
I would not know how. I think users should just use the existing checkpointing utilities if they want to make use of this.